File: src/padflies/padflies/actor.py
from rclpy.node import Node
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from geometry_msgs.msg import PoseStamped

from ._hl_commander_minimal import HighLevelCommander
from ._ll_commander_minimal import LowLevelCommander

from crazyflies.safe.safe_commander import SafeCommander
from .yaw_commander import YawCommander
from ._padflie_tf import PadflieTF


from enum import Enum, auto
from typing import List, Callable, Optional, Tuple

from collision_avoidance_interfaces.srv import CollisionAvoidance
import re
import numpy as np


class ActorState(Enum):
    DEACTIVATED = auto()
    LOW_LEVEL_COMMANDER = auto()
    HIGH_LEVEL_COMMANDER = auto()
    ERROR_STATE = auto()


class PadflieActor:
    def __init__(
        self,
        node: Node,
        tf_manager: PadflieTF,
        hl_commander: HighLevelCommander,
        ll_commander: LowLevelCommander,
        sleep: Callable[[float], None],
    ):
        self._node = node
        self._tf_manager = tf_manager
        self._hl_commander = hl_commander
        self._ll_commander = ll_commander
        self._sleep = sleep

        ### Instance variables
        self._state: ActorState = ActorState.DEACTIVATED
        self._fixed_yaw: bool = False

        self._control_rate: float = 10.0  # Hz
        self._dt: float = 1 / self._control_rate  # Loop period for LL Commander

        ### Commanders (non ROS)
        self.max_rotational_speed: float = 0.5  # Dependent on update rate??
        self.max_step_distance_xy: float = 3  # in meters/second
        self.max_step_distance_z: float = 1  # in meters/second
        self.clipping_box: Optional[List[float]] = None
        self.fixed_yaw_target: float = 0.0

        # TODO: This is from old isse_core for a stable flight
        self.target_history_size: int = 13
        self.__target_history: "list[list[float]]" = []

        self._yaw_commander = YawCommander(
            dt=self._dt,
            max_rotational_speed=self.max_rotational_speed,
        )
        self._safe_commander = SafeCommander(
            dt=self._dt,
            max_step_distance_xy=self.max_step_distance_xy,
            max_step_distance_z=self.max_step_distance_z,
            clipping_box=self.clipping_box,
        )

        # Control Variables, These should not be changed from outside.
        self.__target_pose: Optional[PoseStamped] = None
        self.__last_valid_target_position_and_yaw: Optional[
            Tuple[List[float], float]
        ] = None
        self.__current_yaw: float = (
            0.0  # We dont get yaw from cf. Assume this gets correctly tracked across flight
        )

    def activate(self):
        self._node.get_logger().info("Creating Publishers")
        self._hl_commander.create_publishers()
        self._ll_commander.create_publishers()

        self.__cmd_timer = self._node.create_timer(
            timer_period_sec=self._dt,
            callback=self.__send_target,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )  # This timer needs to be executed while we takeoff or land -> different callback group

        self._collision_avoidance_client = self._node.create_client(
            CollisionAvoidance,
            "collision_avoidance",
            callback_group=MutuallyExclusiveCallbackGroup(),
        )

        self._state = ActorState.HIGH_LEVEL_COMMANDER

    def deactivate(self):
        self._node.destroy_timer(self.__cmd_timer)
        if self._state == ActorState.LOW_LEVEL_COMMANDER:
            self._hl_commander.land(0.0, 5.0)  # Failing semi safe
        self._hl_commander.destroy_publishers()
        self._ll_commander.destroy_publishers()
        self._node.destroy_client(self._collision_avoidance_client)

    def __calculate_collision_avoidance_target(
        self, position: "list[float]", target: "list[float]"
    ) -> Optional["list[float]"]:
        req = CollisionAvoidance.Request()
        req.id = int(re.search(r"\d+", self._hl_commander._prefix).group())
        req.position.x, req.position.y, req.position.z = position
        req.target.x, req.target.y, req.target.z = target
        fut = self._collision_avoidance_client.call_async(req)

        for i in range(10):
            if fut.done():
                break
            self._sleep(0.01)
        if fut.result() is None:
            return None
        res = fut.result()
        return [res.target.x, res.target.y, res.target.z]

    def __send_target(self):
        """Callback to the LOW level commander timer.

        Sends low level commands to the crazyflie if:
                In LOW level commander
            and A target is set

        This is the case after a takeoff.
        If there is no valid crazyflie position we failsafe.
        If there is no transform to the target pose available we choose the last valid one.
        If there was no last valid target position we choose crazyflie position in order to hover.

        After a target is choosen it passes through safe_commanders in order not to crash the crazyflie.
        This is because the commanded positions are not checked before sent to kalman onboard crazyflie.
        """
        if (
            self._state is ActorState.LOW_LEVEL_COMMANDER
            and self.__target_pose is not None
        ):
            cf_position = self._tf_manager.get_cf_position()
            if cf_position is None:
                self._fail_safe("No Crazyflie position.")
                return

            target_position_and_yaw = (
                self._tf_manager.pose_stamped_to_world_position_and_yaw(
                    self.__target_pose
                )
            )

            target_position: List[float]
            target_yaw: float
            # Get target position and target yaw either from our target,
            #                                           the last valid target
            #                                           or the cf position
            if target_position_and_yaw is not None:
                # Target is ok.
                target_position, target_yaw = target_position_and_yaw
                self.__last_valid_target_position_and_yaw = target_position_and_yaw
            else:
                # Target not ok.
                # TODO: Should we hover or should we follow last valid target??
                if self.__last_valid_target_position_and_yaw is not None:
                    target_position, target_yaw = (
                        self.__last_valid_target_position_and_yaw
                    )
                else:
                    # Lets hover in place. This case is probably rare.
                    target_position = cf_position
                    target_yaw = (
                        self.fixed_yaw_target
                    )  # This also might indicate to user that something is not right

            if self._fixed_yaw:
                target_yaw = self.fixed_yaw_target

            # Collision avoidance.
            col_av_target_position = self.__calculate_collision_avoidance_target(
                cf_position, target_position
            )
            if col_av_target_position is not None:
                target_position = col_av_target_position

            # Average out targets (from old ISSE_CORE)
            self.__target_history.insert(0, target_position)
            self.__target_history = self.__target_history[: self.target_history_size]

            ln = len(self.__target_history)
            target_position = list(
                np.average(
                    self.__target_history,
                    axis=0,
                    weights=[((ln - i) / float(ln)) ** 2.6 for i in range(ln)],
                )
            )

            # Pass through safe commanders.
            safe_target = self._safe_commander.safe_cmd_position(
                cf_position, target_position
            )
            safe_yaw = self._yaw_commander.safe_cmd_yaw(self.__current_yaw, target_yaw)

            if self._state is ActorState.LOW_LEVEL_COMMANDER:
                # Due to timing this might happen.
                self._ll_commander.cmd_position(safe_target, safe_yaw)

            # Finally set our current yaw to the yaw we commanded
            # TODO: Is max_yawrate set low enough for this to be safe?
            self.__current_yaw = safe_yaw

    def _fail_safe(self, msg: str):
        self._state = ActorState.ERROR_STATE
        self._hl_commander.land(0.0, 5.0)
        self._node.get_logger().info(f"Transitioning to Error state. {msg}")

    def get_target_pose(self) -> Optional[PoseStamped]:
        return self.__target_pose

    def get_yaw(self):
        return self.__current_yaw

    def set_target(self, target: PoseStamped, use_yaw: bool):
        self.__target_pose = target
        self._fixed_yaw = not use_yaw

    def takeoff_routine(self, takeoff_height: float = 1.0) -> bool:
        """Routine for TAKEOFF.

        Takes off and puts this actor in LowLevel commander mode.
        This ensures that as soon as th crazyflie is in the air it is commanded with cmd_positions.

        Precondition: Crazyflie is in its Pad.
        After: Crazyflie hovers at takeoff_height above the pad. We are in LOW level commander Mode.

        Args:
            position (List[float]): The position we are currently at.
            takeoff_height (float, optional): We will idle at this height above the position after taking off. Defaults to 1.0.
        """
        if self._state is ActorState.ERROR_STATE:
            return False

        self._state = ActorState.HIGH_LEVEL_COMMANDER
        target_pose: Optional[PoseStamped] = self._tf_manager.get_pad_pose_world()
        if target_pose is None:
            return False
        target_pose.pose.position.z += takeoff_height

        self.set_target(
            target_pose, use_yaw=False
        )  # Set target so crazyflie hovers after takeoff

        self._hl_commander.go_to(
            x=0.0, y=0.0, z=0.1, yaw=0.0, duration_seconds=1.0, relative=True
        )
        self._sleep(0.5)
        self._hl_commander.go_to(
            x=0.0, y=0.0, z=0.6, yaw=0.0, duration_seconds=4, relative=True
        )

        # Even though we specified more time for takeoff, this ensures a cleaner transition.
        self._sleep(2)
        self._state = ActorState.LOW_LEVEL_COMMANDER
        return True

    def land_routine(self) -> bool:
        """Routine for LAND.

        Lands the crazyflie safely into the pad.

        Precondition: Crazyflie is in the air, not too far from the Pad (max. 0.75 Meters)
        After: Crazyflie is seated in the pad. Motors are off. State is HIGH Level Commander
        """
        if self._state is ActorState.ERROR_STATE:
            return

        self._ll_commander.notify_setpoints_stop(200)
        """
        Phase2: 
            We are now only using HighLevelCommander
            Use GoTos to properly land. 
            We were 0.5 meters above the pad. 
            First Step: lower to 0.2 m above pad. This is because Phase1 probably overshoots
            Second Step: Fly straight down into the pad. For a good seating.

            During this phase a moving pad is bad but we need the clean high level command routines for safe flight. 

            Between each command the pad position gets querried again.
        """
        p_p_and_yaw = self._tf_manager.get_pad_position_and_yaw()

        self._state = ActorState.HIGH_LEVEL_COMMANDER
        if p_p_and_yaw is None:
            self._fail_safe("No pad position found for landing (Phase2a).")
            return False

        pad_position, yaw = p_p_and_yaw
        self._hl_commander.go_to(
            pad_position[0],
            pad_position[1],
            pad_position[2] + 0.25,
            yaw=yaw,
            duration_seconds=4.5,
        )
        self._sleep(4.5)  # Bleed of momentum we still have from target flight.

        p_p_and_yaw = self._tf_manager.get_pad_position_and_yaw()
        if p_p_and_yaw is None:
            self._fail_safe("No pad position found for landing (Phase2b).")
            return False

        pad_position, yaw = p_p_and_yaw
        self._hl_commander.go_to(
            pad_position[0],
            pad_position[1],
            pad_position[2] - 0.1,
            yaw=yaw,
            duration_seconds=5.0,
        )
        self._sleep(5)  # The actual landing.

        """
        Phase3: 
            We are already in the Pad. 
            We land to stop motors and this Wiggles us to properly seat into the pad.
        """
        p_p_and_yaw = self._tf_manager.get_pad_position_and_yaw()
        if p_p_and_yaw is None:
            self._fail_safe("No pad position found for landing (Phase3).")
            return False

        pad_position, yaw = p_p_and_yaw
        self._hl_commander.land(
            target_height=-0.5, duration_seconds=2.5, yaw=yaw
        )  # Landing at height 0 ensures motors get shut off completely
        self._sleep(0.5)
        # Its ok to  launch again now, but landing rights need to be released!!!
        return True

File: src/padflies_cpp/src/actor.cpp
#include "padflies_cpp/actor.hpp"

static std::unordered_map<std::string, rclcpp::CallbackGroup::SharedPtr> m_callback_groups;
// https://github.com/ros2/rclcpp/pull/2683/commits/86d831375e8a7acdc55272866e04f4c214002414
// As soon as we switch to jazzy or newer we can make this a member variable, currently it would segfault on deconstruction

PadflieActor::PadflieActor(
    std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node,
    const std::string & cf_prefix, 
    PadflieTF* padflie_tf
)
: m_state(ActorState::DEACTIVATED)
, m_dt(0.1) // Default time step
, m_target_pose()
, m_fixed_yaw(false)
, m_yaw_controller(m_dt, 0.5) // Default max rotational velocity of 0.5 rad/s
, m_position_controller(m_dt, 3.0, 1.5, { 3.5, 4.0, 4.500, -7.5, -2.0, 0.0 }) // Default clipping box
, m_collision_avoidance_client(node, std::stoi(cf_prefix.substr(3))) // Extract ID from cf_prefix (/cfID)
, m_hl_commander(node, cf_prefix)
, m_ll_commander(node, cf_prefix)
, m_padflie_tf(padflie_tf)
{
    if (m_callback_groups.find(cf_prefix) == m_callback_groups.end())
        m_callback_groups[cf_prefix] = node->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

    m_send_target_timer = node->create_wall_timer(
        std::chrono::milliseconds((long int)m_dt * 1000),
        std::bind(&PadflieActor::m_send_target_callback, this),
        m_callback_groups[cf_prefix]
    );
}

PadflieActor::~PadflieActor()
{
    //m_send_target_timer.reset(); // Reset the timer to stop it
    RCLCPP_INFO(rclcpp::get_logger("PadflieActor"), "PadflieActor destructor called for ");
}

void PadflieActor::set_target(
    const geometry_msgs::msg::PoseStamped & target_pose,
    bool use_yaw)
{
    m_target_pose = target_pose;
    m_fixed_yaw = !use_yaw;
}

void PadflieActor::get_target_pose(
    geometry_msgs::msg::PoseStamped & target_pose) const
{
    target_pose = m_target_pose;
}

bool PadflieActor::takeoff_routine(
    double takeoff_height)
{
    if (m_state == ActorState::ERROR_STATE)
        return false;
    
    m_state = ActorState::HIGH_LEVEL_COMMANDER;
    geometry_msgs::msg::PoseStamped target_pose;
    if (!m_padflie_tf->get_pad_pose_world(target_pose))
    {
        RCLCPP_ERROR(rclcpp::get_logger("PadflieActor"), "Aborting takeoff, position not available.");
        return false;
    }

    target_pose.pose.position.z += takeoff_height;
    this->set_target(target_pose, false); // Set target so that after takeoff we will hover.

    m_hl_commander.go_to(
        Eigen::Vector3d(0,0,0.1), // relative position to current position
        0.0,                      // yaw
        1.0,                      // duration in seconds
        true);                    // relative movement

    std::this_thread::sleep_for(std::chrono::milliseconds(500));
 
    m_hl_commander.go_to(
        Eigen::Vector3d(0,0,0.6), // relative position to current position
        0.0,                      // yaw
        4.0,                      // duration in seconds
        true);                    // relative movement

    std::this_thread::sleep_for(std::chrono::milliseconds(2000));

    // Even though we specified more time for takeoff, this ensures a cleaner transition.
    m_state = ActorState::LOW_LEVEL_COMMANDER;
    return true; // Indicate successful takeoff
}

/** Routine for LAND.
 *  Lands the crazyflie safely into the pad.
 *
 *  Precondition: Crazyflie is in the air, not too far from the Pad (max. 0.75 Meters)
 *  After: Crazyflie is seated in the pad. Motors are off. State is HIGH Level Commander
 */     
bool PadflieActor::land_routine()
{
    if (m_state == ActorState::ERROR_STATE)
        return false;
    m_ll_commander.notify_setpoints_stop(200);
    m_state = ActorState::HIGH_LEVEL_COMMANDER;

    /**
     * Phase2: 
     *  We are now only using HighLevelCommander
     *  Use GoTos to properly land. 
     *  We were 0.5 meters above the pad. 
     *  First Step: lower to 0.2 m above pad. This is because Phase1 probably overshoots
     *  Second Step: Fly straight down into the pad. For a good seating.
     *
     *  During this phase a moving pad is bad but we need the clean high level command routines for safe flight. 
     *
     *  Between each command the pad position gets querried again.
    */
    double pad_yaw; 
    Eigen::Vector3d pad_position;

    if (!m_padflie_tf->get_pad_position_and_yaw(pad_position, pad_yaw))
    {
        this->fail_safe("No pad position found for landing (Phase2a).");
        return false;
    }

    m_hl_commander.go_to(
        pad_position + Eigen::Vector3d(0, 0, 0.25),// global position above pad
        pad_yaw,                                   // yaw
        4.5);                                      // duration in seconds

    std::this_thread::sleep_for(std::chrono::milliseconds(4500));

    if (!m_padflie_tf->get_pad_position_and_yaw(pad_position, pad_yaw))
    {
        this->fail_safe("No pad position found for landing (Phase2b).");
        return false;
    }

    m_hl_commander.go_to(
        pad_position + Eigen::Vector3d(0, 0, -0.1), // global position in pad
        pad_yaw,                                    // yaw
        1.5);                                       // duration in seconds
    
    std::this_thread::sleep_for(std::chrono::milliseconds(5000));

    /**
     * Phase3:
     * We are already in the Pad. 
     * We land to stop the motors and this will also wiggle us to properly seat in the pad.
    */
    if (!m_padflie_tf->get_pad_position_and_yaw(pad_position, pad_yaw))
    {
        this->fail_safe("No pad position found for landing (Phase3).");
        return false;
    }

    m_hl_commander.land(
        -0.5,       // target height
        2.5,        // duration in seconds
        pad_yaw);   // yaw
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    return true; // Indicate successful landing
}

double PadflieActor::get_yaw() const
{
    return m_current_yaw;
}
std::string PadflieActor::get_current_target_frame() const
{
    return m_target_pose.header.frame_id;
}


void PadflieActor::m_send_target_callback()
{
    if (m_state == ActorState::LOW_LEVEL_COMMANDER)
    {
        Eigen::Vector3d position;
        if (!m_padflie_tf->get_cf_position(position))
        {
            this->fail_safe("Failed to get current position for sending target.");
            return;
        }

        Eigen::Vector3d target_position;
        double target_yaw;
        if (!m_padflie_tf->pose_stamped_to_world_position_and_yaw(m_target_pose, target_position, target_yaw))
        {
            // Target is not ok. Take last valid target. (Hover)
            // Or take cf position (which is also a hover)
            if (m_last_target_valid)
            {
                target_position = m_last_valid_target_position;
                target_yaw = m_last_valid_target_yaw;
            }
            else
            {
                target_position = position; // Use current position as target
                target_yaw = m_fixed_yaw_target; // Use current yaw
            }
        }

        if (m_fixed_yaw) target_yaw = m_fixed_yaw_target;

        m_collision_avoidance_client.get_collision_avoidance_target(position, target_position);

        Eigen::Vector3d safe_target;
        m_position_controller.safe_command_position(position, target_position, safe_target);
        double safe_yaw = m_yaw_controller.safe_cmd_yaw(m_current_yaw, target_yaw);

        // This is for race conditions and should be removed if possible.
        if (m_state == ActorState::LOW_LEVEL_COMMANDER)
        {
            m_ll_commander.cmd_position(safe_target, safe_yaw);
        }
        
        m_current_yaw = safe_yaw; 
    }
}

void PadflieActor::fail_safe(std::string reason)
{
    m_state = ActorState::ERROR_STATE;
    m_hl_commander.land(
        -0.5,       // target height
        4.0,        // duration in seconds
        0.0);       // yaw
    RCLCPP_ERROR(rclcpp::get_logger("PadflieActor"), "Fail-safe triggered! Landing in place. %s", reason.c_str());
}